### ISSUE TITLE
Improve ParallelDatasetIterator documentation

### ISSUE THREAD

**Issue by trypag** [torchnet/torchnet#70]
Hey,
This is not really an issue, I would like to have so more details about how this iterator operates the threading. I came through situations where the original dataset was loaded `n` times, where `n` is the number of threads. What I would like to do is load my dataset on-the-fly and let torchnet do the multitasking. Problem here is because I don't understand how and what torchnet will multitask, I can't write my iterator efficiently.

I tried to get my hands into the code but it's really hard to understand how `Threads` operates. I hope someone will find the time to answer this question. Maybe we can improve the documentation to make it less error-prone 😃

**Comment by trypag** [comment]
Here is a sample from the mnist.lua example, it could be nice to show how it can be used with more than 1 thread, maybe another dataset could be use, this is not what is important. Showing what will be multi-threaded with how much data would be interesting.

```lua
local function getIterator(mode)
   return tnt.ParallelDatasetIterator{
      nthread = 1, -- this is not really multi-threaded
      init    = function() require 'torchnet' end,
      closure = function()

         -- load MNIST dataset:
         local mnist = require 'mnist'
         local dataset = mnist[mode .. 'dataset']()
         dataset.data = dataset.data:reshape(dataset.data:size(1),
            dataset.data:size(2) * dataset.data:size(3)):double()

         -- return batches of data:
         return tnt.BatchDataset{
            batchsize = 128,
            dataset = tnt.ListDataset{  -- replace this by your own dataset
               list = torch.range(1, dataset.data:size(1)):long(),
               load = function(idx)
                  return {
                     input  = dataset.data[idx],
                     target = torch.LongTensor{dataset.label[idx] + 1},
                  }  -- sample contains input and target
               end,
            }
         }
      end,
   }
end
```

**Comment by trypag** [comment]
Another simplified example from https://github.com/karandwivedi42/imagenet-multiGPU.torchnet/blob/master/main.lua. This one is multi-threaded, from what I see each thread will receive the closure and execute it, it will yield that each thread will have produced the same `Dataset`. I don't get where the data is separated on each thread and makes it really suitable for multi-threaded jobs (ie each thread does the same job on different datasets).

 ```lua
local function getIterator(mode)
   local iterator = tnt.ParallelDatasetIterator{
      nthread = opt.nDonkeys,
      init = function(nthread)
         require 'torchnet'
         require 'randomdataset'
         require 'image'
         transforms = require 'transforms'
      end,
      closure = function()
         local dataset = provider[mode]
         local classes = provider.classes
         local classlist= {}
         for l= 1,#dataset do
            classlist[#classlist+1] = tnt.ListDataset{
               list = dataset[l],
               load = function(im)
                  return {
                     input = image.load(paths.concat(opt.data,mode,classes[l],im),3,'float'),
                     target = torch.LongTensor{l}
                  }
               end,
            }
         end
         
         local concat =  tnt.ConcatDataset{datasets = classlist}
         -- keep ordering same in val, in train acts as class selector
         if mode == 'train' then concat = concat:random() end

         return concat:transform{
            input =
               mode == 'train' and
                  tnt.transform.compose{
                     transforms.Scale(opt.imageSize),
                     transforms.RandomCrop(opt.cropSize),
                     transforms.ColorNormalize(provider.meanstd),
                     transforms.HorizontalFlip(0.5),
                  }
              or
                  tnt.transform.compose{
                     transforms.Scale(opt.imageSize),
                     transforms.CenterCrop(opt.cropSize),
                     transforms.ColorNormalize(provider.meanstd),
                  }
               }:batch(opt.batchSize,'skip-last')
      end,
   }
   return iterator
end
```

**Comment by karandwivedi42** [comment]
@trypag From my limited understanding, each thread will have a different Dataset object for itself. (The dataset objects may share their internal tensors because of `sharedserialize`).

> What I would like to do is load my dataset on-the-fly and let torchnet do the multitasking. 

For this I created a `provider` *outside* the threads to collect all information from the folders only once. Then the threads only do the image loading and preprocessing.

You can also see https://github.com/soumith/imagenet-multiGPU.torch for an implementation of the same dataloader using Threads without torchnet.

**Comment by trypag** [comment]
@karandwivedi42 thanks for your help, your implementation was really helpful to understand torchnet 😃 
I understand the trick related to the `provider` a few days ago, it was not clear at first. What I don't get though is when the data starts to be "divided' for each thread.

For example, I think this part of your code is going to be executed `nDonkeys` time :
```lua
        for l= 1,#dataset do
           classlist[#classlist+1] = tnt.ListDataset{
              list = dataset[l],
              load = function(im)
                 return {
                    input = image.load(paths.concat(opt.data,mode,classes[l],im),3,'float'),
                    target = torch.LongTensor{l}
                 }
              end,
           }
        end
```
This will result in having `nDonkeys` identical classlist tables, one on each thread.

**Comment by lvdmaaten** [comment]
Suppose you do something like the following:
```
local tds = require 'tds'
local tnt = require 'torchnet'
local imagelist = tds.Vec{'image1.jpg', 'image2.jpg'}
local iterator = tnt.ParallelDatasetIterator{
   nthreads = 10,
   init     = function() require 'torchnet' end,
   closure  = function()
      return tnt.ListDataset{
         list = imagelist,
         load = function(filename)
            require 'image'
            return {input = image.load(filename)}
         end
      }
   end
}
```
Each thread has its own copy of the `ListDataset`. However, note that this dataset hardly contains anything (just a few references) so this is perfectly fine. In fact, since we're using `sharedserialize` we are not even making a copy of `imagelist` at all.

The `ParallelDatasetIterator` will call `:get(idx)` on the `ListDataset`, for instance, it will call `get(1)` on the dataset in thread 1 and `get(2)` on the dataset in thread 2. If you look at the implementation of `ListDataset.get()`, you see that this invokes the load closure. So `'image1.jpg'` will only be loaded by a single thread, and `'image2.jpg'` will only be loaded by a single (potentially different) thread. This is exactly what you want.

**Comment by lvdmaaten** [comment]
Note that in the MNIST example, also, there is no cloning of data in threads because the code uses `sharedserialize` on `dataset.data` and `dataset.label`, which means that all the threads have references to the same memory.

@karandwivedi42: The `imagenet-multiGPU.torch` is pretty awful because all the abstractions leak into each other there. In fact, it was one of the motivations for us to develop torchnet.

**Comment by karandwivedi42** [comment]
@lvdmaaten thanks!

@trypag Yes, that loop will be executed on each thread.
> This will result in having nDonkeys identical classlist tables, one on each thread
True

**Comment by trypag** [comment]
@lvdmaaten Thank you for this example, it gives more understanding on how `ParallelDatasetIterator` should be used. I think I got it ! Please don't close this issue yet 😄 . I will write another point related to this in a few hours, and hopefully get your expertise. thank you @karandwivedi42 !

**Comment by trypag** [comment]
1. @lvdmaaten First question related to your previous comment :
> Each thread has its own copy of the ListDataset. However, note that this dataset hardly contains anything (just a few references) so this is perfectly fine. In fact, since we're using sharedserialize we are not even making a copy of imagelist at all.

When the `Iterator` is instantiated, each thread starts the closure with a shared `imagelist`, and builds a `ListDataset`. Meaning they all have their own `ListDataset` in memory (not shared) with the same values in it ? Maybe this is required to do so, I just want to be sure how it works.

2. You said the thread are using `sharedserialize`, does it mean that once each thread has executed his closure, the returned `Dataset` is serialized on the disk, instead of keeping it in memory ?

Thanks !

**Comment by trypag** [comment]
#### Extracting patches from images on-the-fly
------------------------------------------------------------------------------------------------------------------------------------------------
FYI : this is a public research application on medical imaging

We want to solve a segmentation problem at the pixel scale, meaning that we need to predict one class for each pixel in the image. For doing so, one approach is to extract patches in the image and associate the label of the center pixel to it. Here is what I am doing :
- First build a `ListDataset` of images, this is simple !
- For each images of this `Dataset` build a new `Dataset` of patches. My current implementation is extracting patches' center position (so not accessing to the image data).
- Then when calling `get(idx)` I should return a sub-view of the image stored on the GPU. **This is the point where I am not sure about** !!!

What I would want is to have the minimum number of image (16 Mo) on the GPU while returning my patches. The idea behing this is to lower the cost of transfers, maybe this is over-optimization. I think this is also a general question : How should I load my patches so that the GPU is not overwelmed in memory and the transfer between cpu and gpu low. I had in mind to copy each batch of patches to the GPU, maybe that would be inefficient. I think this could be transposed in the same way to loading the patches by keeping the minimum images in memory.

**The challenge here is that it is multi-threaded and I want to load the minimum image required to extract the queried patches**

I understand that maybe this is not possible to implement this.

**Comment by lvdmaaten** [comment]
Yeah, this is tricky because the `ParallelDatasetIterator` may allocate large amounts of memory on the GPU, where memory is sparse.

In general, I use an approach where I do all the data preprocessing (such as the cropping you describe, but also the batching, etc) on the CPU-side. So I have my `ParallelDatasetIterator` spit out batches of images in CPU memory that are ready to be fed into the network. Then, I copy the batch to the CPU in the `engine.hooks.onSample()` hook (that is, right before the call to `network:forward()`). This has the added benefit that it makes it really easy to re-use the same GPU memory buffer for every batch --- this means I'm never allocating or freeing GPU memory during training (a GPU free is inefficient because it adds a synchronization point).

**Comment by trypag** [comment]
@lvdmaaten This is smart ! really nice that you took the time to help me, thank you. 
Do you think we should sum up this discussion in some kind of documentation file ? maybe doc/ParralelDatasetIterator.md ?
I would be happy to give back. If you think this is not necessary, it's ok for me also.

**Comment by karandwivedi42** [comment]
@trypag you can see an example of
> re-use the same GPU memory buffer for every batch

here: https://github.com/karandwivedi42/imagenet-multiGPU.torchnet/blob/master/main.lua#L232-L233

Also, to answer the question about cropping, then you can do that within the `ListDataset`. When you return the image and label [here](https://github.com/karandwivedi42/imagenet-multiGPU.torchnet/blob/master/main.lua#L170) you can instead return a random crop (or centre crop). To view each image more than once, you can use a `ResampleDataset`. The disadvantage is that same image will be loaded from disk more than once, but if you are loading small images from SSD, it should be very fast and almost consumed in the time of forward match.
Just remember to keep separate randomSeeds for each thread if you are using `random` anywhere in the iterator.

**Comment by lvdmaaten** [comment]
Even better than putting everything in `load()` would be to wrap your `ListDataset` into a `TransformDataset` that implements the image transforms. This makes it easy to plug in a different source of images later when needed.
